<a href="https://colab.research.google.com/github/choihana/opentutorials_deeplearning/blob/master/%EC%95%84%EC%9D%B4%EB%A6%AC%EC%8A%A4_%ED%92%88%EC%A2%85_%EB%B6%84%EB%A5%98%ED%95%98%EA%B8%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
# 라이브러리 사용
import tensorflow as tf
import pandas as pd

In [4]:
# 데이터 준비
file1 = 'https://raw.githubusercontent.com/blackdew/tensorflow1/master/csv/iris.csv'
iris = pd.read_csv(file1)
iris.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭      품종
0   5.1  3.5    1.4   0.2  setosa
1   4.9  3.0    1.4   0.2  setosa
2   4.7  3.2    1.3   0.2  setosa
3   4.6  3.1    1.5   0.2  setosa
4   5.0  3.6    1.4   0.2  setosa

In [5]:
#원핫인코딩으로 품종 0,1로 바꿔주기

In [6]:
encoded = pd.get_dummies(iris)

In [9]:
encoded.head()

꽃잎길이  꽃잎폭  꽃받침길이  꽃받침폭  품종_setosa  품종_versicolor  품종_virginica
0   5.1  3.5    1.4   0.2          1              0             0
1   4.9  3.0    1.4   0.2          1              0             0
2   4.7  3.2    1.3   0.2          1              0             0
3   4.6  3.1    1.5   0.2          1              0             0
4   5.0  3.6    1.4   0.2          1              0             0

In [10]:
encoded.columns

Index(['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭', '품종_setosa', '품종_versicolor',
       '품종_virginica'],
      dtype='object')

In [11]:
# feature, target 분류
feature = encoded[['꽃잎길이', '꽃잎폭', '꽃받침길이', '꽃받침폭']]
target = encoded[['품종_setosa', '품종_versicolor',
       '품종_virginica']]
print(feature.shape,target.shape)

(150, 4) (150, 3)


In [12]:
# 모델 구조 만들기
X = tf.keras.layers.Input(shape=[4])
y = tf.keras.layers.Dense(3, activation='softmax')(X)
model = tf.keras.models.Model(X,y)
model.compile(loss='categorical_crossentropy', metrics='accuracy')

In [13]:
# 데이터로 모델 학습 시키기
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 4.8447 - accuracy: 0.3333
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 4.6577 - accuracy: 0.3333
Epoch 3/10
5/5 [==============================] - 0s 1ms/step - loss: 4.5202 - accuracy: 0.3333
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 4.3985 - accuracy: 0.3333
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 4.2830 - accuracy: 0.3333
Epoch 6/10
5/5 [==============================] - 0s 3ms/step - loss: 4.1714 - accuracy: 0.3333
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 4.0625 - accuracy: 0.3333
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 3.9536 - accuracy: 0.3333
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 3.8431 - accuracy: 0.3333
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 3.7355 - accuracy: 0.3333


In [14]:
model.fit(feature, target, epochs=100, verbose=0)

In [15]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4344 - accuracy: 0.8733
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4324 - accuracy: 0.8867
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4315 - accuracy: 0.8867
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4299 - accuracy: 0.8867
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4290 - accuracy: 0.8933
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4273 - accuracy: 0.8933
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4281 - accuracy: 0.9000
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4250 - accuracy: 0.8800
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.4240 - accuracy: 0.8867
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 0.4234 - accuracy: 0.8867


In [16]:
model.fit(feature, target, epochs=100,verbose=0)

In [17]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3404 - accuracy: 0.9400
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3391 - accuracy: 0.9400
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3386 - accuracy: 0.9400
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3389 - accuracy: 0.9333
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3370 - accuracy: 0.9400
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3369 - accuracy: 0.9400
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3387 - accuracy: 0.9333
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3353 - accuracy: 0.9400
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3367 - accuracy: 0.9333
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3343 - accuracy: 0.9400


In [18]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3342 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3333 - accuracy: 0.9400
Epoch 3/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3328 - accuracy: 0.9400
Epoch 4/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3327 - accuracy: 0.9400
Epoch 5/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3323 - accuracy: 0.9333
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3317 - accuracy: 0.9400
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3309 - accuracy: 0.9400
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3300 - accuracy: 0.9400
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3293 - accuracy: 0.9400
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3287 - accuracy: 0.9400


In [19]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3278 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3287 - accuracy: 0.9333
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3273 - accuracy: 0.9467
Epoch 4/10
5/5 [==============================] - 0s 3ms/step - loss: 0.3263 - accuracy: 0.9467
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3261 - accuracy: 0.9467
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3256 - accuracy: 0.9400
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3245 - accuracy: 0.9467
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3246 - accuracy: 0.9400
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3235 - accuracy: 0.9333
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3229 - accuracy: 0.9400


In [20]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3229 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3217 - accuracy: 0.9400
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3211 - accuracy: 0.9400
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3208 - accuracy: 0.9467
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3207 - accuracy: 0.9400
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3195 - accuracy: 0.9467
Epoch 7/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3198 - accuracy: 0.9400
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3189 - accuracy: 0.9467
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3183 - accuracy: 0.9400
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3176 - accuracy: 0.9400


In [21]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3174 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3165 - accuracy: 0.9467
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3156 - accuracy: 0.9467
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3161 - accuracy: 0.9400
Epoch 5/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3156 - accuracy: 0.9467
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3148 - accuracy: 0.9400
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3136 - accuracy: 0.9467
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3132 - accuracy: 0.9467
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3131 - accuracy: 0.9467
Epoch 10/10
5/5 [==============================] - 0s 3ms/step - loss: 0.3120 - accuracy: 0.9467


In [22]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3124 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3113 - accuracy: 0.9533
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3105 - accuracy: 0.9533
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3113 - accuracy: 0.9467
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3099 - accuracy: 0.9533
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3105 - accuracy: 0.9533
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3087 - accuracy: 0.9533
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3088 - accuracy: 0.9533
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3089 - accuracy: 0.9533
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3076 - accuracy: 0.9467


In [23]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3077 - accuracy: 0.9533
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3065 - accuracy: 0.9467
Epoch 3/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3071 - accuracy: 0.9467
Epoch 4/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3060 - accuracy: 0.9400
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3058 - accuracy: 0.9467
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3047 - accuracy: 0.9467
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3044 - accuracy: 0.9533
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3041 - accuracy: 0.9600
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3032 - accuracy: 0.9533
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3032 - accuracy: 0.9467


In [24]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3027 - accuracy: 0.9467
Epoch 2/10
5/5 [==============================] - 0s 1ms/step - loss: 0.3027 - accuracy: 0.9467
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3014 - accuracy: 0.9533
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3019 - accuracy: 0.9533
Epoch 5/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3003 - accuracy: 0.9533
Epoch 6/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3001 - accuracy: 0.9533
Epoch 7/10
5/5 [==============================] - 0s 2ms/step - loss: 0.3001 - accuracy: 0.9467
Epoch 8/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2999 - accuracy: 0.9533
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2984 - accuracy: 0.9533
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2986 - accuracy: 0.9533


In [25]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2982 - accuracy: 0.9533
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2972 - accuracy: 0.9467
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2969 - accuracy: 0.9600
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2965 - accuracy: 0.9467
Epoch 5/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2963 - accuracy: 0.9533
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2963 - accuracy: 0.9533
Epoch 7/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2954 - accuracy: 0.9533
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2946 - accuracy: 0.9533
Epoch 9/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2949 - accuracy: 0.9600
Epoch 10/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2948 - accuracy: 0.9600


In [26]:
# 계속 loss가 낮아지고 accuracy 증가하고 있어서 홀린듯이 계속 학습시키고 있다.. 언제 멈출것인가

In [27]:
model.fit(feature, target, epochs=10)

Epoch 1/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2932 - accuracy: 0.9600
Epoch 2/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2931 - accuracy: 0.9533
Epoch 3/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2938 - accuracy: 0.9467
Epoch 4/10
5/5 [==============================] - 0s 2ms/step - loss: 0.2922 - accuracy: 0.9533
Epoch 5/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2918 - accuracy: 0.9600
Epoch 6/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2918 - accuracy: 0.9533
Epoch 7/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2908 - accuracy: 0.9533
Epoch 8/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2909 - accuracy: 0.9467
Epoch 9/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2899 - accuracy: 0.9533
Epoch 10/10
5/5 [==============================] - 0s 1ms/step - loss: 0.2898 - accuracy: 0.9533


In [28]:
# 이제 멈춰보자

In [31]:
# 모델 이용하기
model.predict(feature[-5:])

array([[7.7685138e-04, 2.9600722e-01, 7.0321596e-01],
       [1.0364646e-03, 4.0508336e-01, 5.9388012e-01],
       [1.4943638e-03, 3.3447605e-01, 6.6402966e-01],
       [6.5177929e-04, 1.2850676e-01, 8.7084150e-01],
       [2.0206724e-03, 2.5294507e-01, 7.4503422e-01]], dtype=float32)

In [32]:
target[-5:]

품종_setosa  품종_versicolor  품종_virginica
145          0              0             1
146          0              0             1
147          0              0             1
148          0              0             1
149          0              0             1

In [33]:
# 학습한 가중치
model.get_weights()

[array([[ 0.709742  ,  0.91348004, -0.4216438 ],
        [ 1.3727269 , -0.6765589 , -0.26938015],
        [-1.1322262 ,  0.18139009,  1.2702117 ],
        [-1.5633103 ,  0.0742832 ,  1.5627613 ]], dtype=float32),
 array([ 0.26999962,  0.3984334 , -0.09786525], dtype=float32)]

1. setosa 인지 예측하는 수식
softmax (0.709742 * x1 + 1.3727269 * x2 + -1.1322262 * x3 + -1.5633103 * x4 + 0.26999962 )

2. versicolor 인지 예측하는 수식
softmax (0.91348004 * x1 + -0.6765589 * x2 + 0.18139009 * x3 + 0.0742832 * x4 + 0.3984334  )

3. virginica 인지 예측하는 수식
softmax (-0.4216438 * x1 + -0.26938015 * x2 + 1.2702117 * x3 +1.5627613 * x4 + -0.09786525)
